QIR specification update 1 (#67)
microsoft/qsharp-language: Specifications/QIR/Callables.md

expected argument tuple; specifically, it can't tell if an inner tuple is the
result of an application of `Controlled` or just part of the base signature.

> ### Implementing Lambdas, Partial Application and Currying

Reviewer: ```suggestion
### Implementing Lambdas, Partial Application. and Currying
```

The lambda body may need to include additional code that performs argument tuple
construction before calling the underlying callable.

#### Memory Management Table

> Any captured values need to remain alive as long as the callable value exists, and correspondingly may also need to be unreferenced when the type information for the captured values cannot be statically determined. Upon creation, a table with two function pointers is hence associated with a callable value. 

Reviewer 2: Maybe, explain here that the memory management is recursive? (Otherwise, the callable could manage the capture tuple automatically and it's not clear why these additional facilities are needed.)

Author: I expanded a bit more here.

> Like the implementation table, the table is defined as global constant with a unique name. It contains two pointers of type `void(%Tuple*, i64)*`; the first one points to the function for modifying the reference counts of captured values, the second points to the one for modifying the access counts. They can be invoked using the runtime function `__quantum__rt__callable_memory_management`. If there are no captured values, a null pointer should be passed upon callable creation.

Reviewer 2: "access counts" -> "alias counts"

Author: good catch

The lambda body may need to include additional code that performs argument tuple
construction before calling the underlying callable.

#### Memory Management Table

Since any captured values need to remain alive as long as the callable value exists, they also need to be unreferenced when the callable value is released. While sufficient type information for the captured values is known upon creation of the value, the information is no longer available at the time when it is released.
Upon creation, a table with two function pointers for modifying reference and alias counts for captured values is hence associated with a callable value. 

> Like the implementation table, the table is defined as global constant with a unique name. It contains two pointers of type `void(%Tuple*, i64)*`; the first one points to the function for modifying the reference counts of captured values, the second points to the one for modifying the alias counts. Either of those pointers may be null, and if no values were captured, a null pointer should be passed instead of a table upon callable creation.

Reviewer: ```suggestion
Like the implementation table, the table is defined as global constant with a unique name. It contains two pointers of type `void(%Tuple*, i64)*`; the first one points to the function for modifying the reference counts of captured values, the second points to the one for modifying the alias counts. Either of those pointers may be null, and if no managed values were captured, a null pointer should be passed instead of a table upon callable creation.
```

I added "managed" because a null pointer should be also provided if all that's captured are some integers or other scalars.

> Like for tuple and array items, the responsibility of managing the reference and access count for captured values lays with the compiler. The two functions can be invoked using the runtime function `__quantum__rt__callable_memory_management`, see the description [below](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Callables.md#runtime-functions). 

Reviewer: ```suggestion
As for tuple and array items, the responsibility of managing the reference and alias count for captured values lays with the compiler. The two functions can be invoked using the runtime function `__quantum__rt__callable_memory_management`, see the description [below](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Callables.md#runtime-functions). 
```

> Like for tuple and array items, the responsibility of managing the reference and access count for captured values lays with the compiler. The two functions can be invoked using the runtime function `__quantum__rt__callable_memory_management`, see the description [below](https://github.com/microsoft/qsharp-language/blob/main/Specifications/QIR/Callables.md#runtime-functions). 

Reviewer: It might be worth giving an example where the ref count or alias count of the capture tuple elements could change without the ref count or access count of the callable changing. The simplest I can come up with is pretty trivial, in pseudo-Q#-ish:
`let f = lambda (x : Int[], y : Int) -> (x, y);`
Then every call to `f` has to increment the ref and access count of the capture tuple elements, specifically of the `x` array.

Author: I assume you meant to have x be captured in this example. This example wouldn't actually require going via the capture update function every time. the only time you really need to update the capture counts is when you update the callable counts. However, neither the compiler not the runtime has enough type information to do that unless we save those functions with the callable, and it just seems more consistent with how other counts work to have count changes always apply only to the exact value given as argument, and not anything "inside" that value.

| __quantum__rt__callable_invoke  | `void(%Callable*, %Tuple*, %Tuple*)` | Invokes the callable with the provided argument tuple and fills in the result tuple. |
| __quantum__rt__callable_make_adjoint | `void(%Callable*)`                         | Updates the callable by applying the Adjoint functor. |
| __quantum__rt__callable_make_controlled | `void(%Callable*)`                      | Updates the callable by applying the Controlled functor. |
| __quantum__rt__callable_update_reference_count | `void(%Callable*, i64)`                      | Adds the given integer value to the reference count for the callable. Deallocates the callable if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |
| __quantum__rt__callable_update_alias_count | `void(%Callable*, i64)`                      | Adds the given integer value to the alias count for the callable. Fails if the count becomes negative. |

> | __quantum__rt__callable_memory_management | `void(i32, %Callable*, i64)`                      | Invokes the function at the given index in the memory management table of the callable with the capture tuple and the given 64-bit integer. Does nothing if if the memory management table pointer or the function pointer at that index is null.  |

Reviewer 2: I'd put %Callable* argument first to be consistent with other methods in the table (and with the convention of passing 'this' pointer) but I don't feel strongly about it. Your call.

Author: Let's go with Alan's suggestion below and actually specify which function has to be at which index and give them distinct names.

| __quantum__rt__callable_invoke  | `void(%Callable*, %Tuple*, %Tuple*)` | Invokes the callable with the provided argument tuple and fills in the result tuple. |
| __quantum__rt__callable_make_adjoint | `void(%Callable*)`                         | Updates the callable by applying the Adjoint functor. |
| __quantum__rt__callable_make_controlled | `void(%Callable*)`                      | Updates the callable by applying the Controlled functor. |
| __quantum__rt__callable_update_reference_count | `void(%Callable*, i64)`                      | Adds the given integer value to the reference count for the callable. Deallocates the callable if the reference count becomes 0. The behavior is undefined if the reference count becomes negative. |
| __quantum__rt__callable_update_alias_count | `void(%Callable*, i64)`                      | Adds the given integer value to the alias count for the callable. Fails if the count becomes negative. |

> | __quantum__rt__callable_memory_management | `void(i32, %Callable*, i64)`                      | Invokes the function at the given index in the memory management table of the callable with the capture tuple and the given 64-bit integer. Does nothing if if the memory management table pointer or the function pointer at that index is null.  |

Reviewer: I would rename this `__quantum__rt__callable_manage_memory`, since generally all of our runtime function names are verbs or verb phrases. I might even split it into two separate functions, `__quantum__rt__callable_update_capture_reference` and `__quantum__rt__callable_update_capture_alias`.

Author: I gave each one separate names.
